Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift

```diff
@@ -258,30 +262,259 @@ internal final class _CalendarGregorian: _CalendarProtocol, @unchecked Sendable
         }
     }
     
+    // Returns the range of a component in Gregorian Calendar.
+    // When there are multiple possible upper bounds, the largest one is returned.
     func maximumRange(of component: Calendar.Component) -> Range<Int>? {
         switch component {
-        case .era: 0..<2
-        case .year: 1..<144684
-        case .month: 1..<13
-        case .day: 1..<32
-        case .hour: 0..<24
-        case .minute: 0..<60
-        case .second: 0..<60
-        case .weekday: 1..<8
-        case .weekdayOrdinal: 1..<6
-        case .quarter: 1..<5
-        case .weekOfMonth: 1..<7
-        case .weekOfYear: 1..<54
-        case .yearForWeekOfYear: 140742..<144684
-        case .nanosecond: 0..<1000000000
-        case .isLeapMonth: 0..<2
+        case .era: return 0..<2
+        case .year: return 1..<144684
+        case .month: return 1..<13
+        case .day: return 1..<32
+        case .hour: return 0..<24
+        case .minute: return 0..<60
+        case .second: return 0..<60
+        case .weekday: return 1..<8
+        case .weekdayOrdinal:
+            return 1..<6
+        case .quarter: return 1..<5
+        case .weekOfMonth:
+            let lowerBound = minimumDaysInFirstWeek == 1 ? 1 : 0
+            let daysInMonthLimit = 31
+            let upperBound = (daysInMonthLimit + 6 + (7 - minimumDaysInFirstWeek)) / 7;
+            return lowerBound ..< (upperBound + 1)
+        case .weekOfYear: return 1..<54
+        case .yearForWeekOfYear: return 140742..<144684
+        case .nanosecond: return 0..<1000000000
+        case .isLeapMonth: return 0..<2
         case .calendar, .timeZone:
-            nil
+            return nil
         }
     }
-    
+
+    // There is a chance of refactoring Calendar_ICU to use these
+    func _algorithmA(smaller: Calendar.Component, larger: Calendar.Component, at: Date) -> Range<Int>? {
+        guard let interval = dateInterval(of: larger, for: at) else {
+            return nil
+        }
+
+        guard let ord1 = ordinality(of: smaller, in: larger, for: interval.start + 0.1) else {
+            return nil
+        }
+
+        guard let ord2 = ordinality(of: smaller, in: larger, for: interval.start + interval.duration - 0.1) else {
+            return nil
+        }
+
+        guard ord2 >= ord1 else {
+            // Protect against an unexpected value from ICU for ord2
```

Comment:
Do we still need this protection here? Maybe if we refactor them together...

Reply:
I removed the comment since that's no longer correct. Although, what refactor are you referring to?

Reply:
Just the eventual possibility that we put this algorithmA,B,C,D stuff into one place. Not a blocker here.

Reply:
Agreed. That's definitely on my plate